bin/hoster/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+		let dht_event_stream =
+			network.event_stream("authority-discovery").filter_map(|e| async move {
+				match e {
+					Event::Dht(e) => Some(e),
+					_ => None,
+				}
+			});
+		let (authority_discovery_worker, _service) =
+			sc_authority_discovery::new_worker_and_service_with_config(
+				sc_authority_discovery::WorkerConfig {
+					publish_non_global_ips: auth_disc_publish_non_global_ips,
+					..Default::default()
+				},
+				client.clone(),
+				network.clone(),
+				Box::pin(dht_event_stream),
+				authority_discovery_role,
+				prometheus_registry.clone(),
+			);
+
+		task_manager
+			.spawn_handle()
+			.spawn("authority-discovery-worker", authority_discovery_worker.run());
+	}
+
+	// if the node isn't actively participating in consensus then it doesn't
+	// need a keystore, regardless of which protocol we use below.
+	let keystore =
+		if role.is_authority() { Some(keystore_container.sync_keystore()) } else { None };
+
+	if enable_grandpa {
+		let config = sc_finality_grandpa::Config {
+			// FIXME #1578 make this available through chainspec
+			gossip_duration: std::time::Duration::from_millis(333),
+			justification_period: 512,
+			name: Some(name),
+			observer_enabled: false,
+			keystore,
+			local_role: role,
+			telemetry: telemetry.as_ref().map(|x| x.handle()),
+		};
+
+		// start the full GRANDPA voter
+		// NOTE: non-authorities could run the GRANDPA observer protocol, but at
+		// this point the full voter should provide better guarantees of block
+		// and vote data availability than the observer. The observer has not
+		// been tested extensively yet and having most nodes in a network run it
+		// could lead to finality stalls.
+		let grandpa_config = sc_finality_grandpa::GrandpaParams {
+			config,
+			link: grandpa_link,
+			network: network.clone(),
+			telemetry: telemetry.as_ref().map(|x| x.handle()),
+			voting_rule: sc_finality_grandpa::VotingRulesBuilder::default().build(),
+			prometheus_registry,
+			shared_voter_state: sc_finality_grandpa::SharedVoterState::empty(),
+		};
+
+		// the GRANDPA voter task is considered infallible, i.e.
+		// if it fails we take down the service with it.
+		task_manager.spawn_essential_handle().spawn_blocking(
+			"grandpa-voter",
+			sc_finality_grandpa::run_grandpa_voter(grandpa_config)?,
+		);
+	}
+
+	network_starter.start_network();
+
+	Ok(task_manager)
+}
+
+#[derive(structopt::StructOpt)]
+struct Cli<GenesisConfig, Extension = sc_chain_spec::NoExtension> {
+	#[structopt(skip)]
+	_phantom: std::marker::PhantomData<(GenesisConfig, Extension)>,
+	#[structopt(flatten)]
+	run: sc_cli::RunCmd,
+}
+
+impl<GenesisConfig, Extension> SubstrateCli for Cli<GenesisConfig, Extension>
+where
+	GenesisConfig: sc_chain_spec::RuntimeGenesis + 'static,
+	Extension:
+		sp_runtime::DeserializeOwned + Send + Sync + sc_service::ChainSpecExtension + 'static,
+{
+	fn impl_name() -> String {
+		"Runtime Hoster".into()
+	}
+
+	fn impl_version() -> String {
+		Default::default()
+	}
```

Comment:
Do we want impl_version to be an empty string?

Reply:
Probably not, but it'll do for now.